README.md
# Django Querycount

![Current Release](http://img.shields.io/pypi/v/django-querycount.svg?style=flat-square) ![MIT License](http://img.shields.io/pypi/l/django-querycount.svg?style=flat-square)

Inspired by [this post by David Szotten](http://goo.gl/UUKN0r), this project
gives you a middleware that prints database query counts in Django's runserver
console output.

![Screenshot](docs/screenshot.png)

## Installation

Install with pip:

```
pip install django-querycount
```

Then add `querycount.middleware.QueryCountMiddleware` to your
`MIDDLEWARE`.

Notice that `django-querycount` only works while your project is in DEBUG mode.

## Settings

There are two possible settings for this app: The first defines threshold
values used to color output, while the second allows you customize requests
that will be ignored by the middleware. The default settings are:

```python
QUERYCOUNT = {
    'THRESHOLDS': {
        'MEDIUM': 50,
        'HIGH': 200,
        'MIN_TIME_TO_LOG':0,
        'MIN_QUERY_COUNT_TO_LOG':0
    },
    'IGNORE_REQUEST_PATTERNS': [],
    'IGNORE_SQL_PATTERNS': [],
    'DISPLAY_DUPLICATES': None,
    'RESPONSE_HEADER': 'X-DjangoQueryCount-Count'
}
```

The `QUERYCOUNT['THRESHOLDS']` settings will determine how many queries are
interpreted as high or medium (and the color-coded output). In previous versions
of this app, this settings was called `QUERYCOUNT_THRESHOLDS` and that setting
is still supported.

The `QUERYCOUNT['IGNORE_REQUEST_PATTERNS']` setting allows you to define a list of
regexp patterns that get applied to each request's path. If there is a match,
the middleware will not be applied to that request. For example, the following
setting would bypass the querycount middleware for all requests to the admin:

```python
QUERYCOUNT = {
    'IGNORE_REQUEST_PATTERNS': [r'^/admin/']
}
```

The `QUERYCOUNT['IGNORE_SQL_PATTERNS']` setting allows you to define a list of
regexp patterns that ignored to statistic sql query count. For example, the following
setting would bypass the querycount middleware for django-silk sql query::

````python
QUERYCOUNT = {
    'IGNORE_SQL_PATTERNS': [r'silk_']
}
```

The `QUERYCOUNT['RESPONSE_HEADER']` setting allows you to define a custom response
header that contains the total number of queries executed. To disable this header,
the supply `None` as the value:

```python
QUERYCOUNT = {
    'RESPONSE_HEADER': None
}
```

*New in 0.4.0*. The `QUERYCOUNT['DISPLAY_DUPLICATES']` setting allows you
to control how the most common duplicate queries are displayed. If the setting
is `None` (the default), duplicate queries are not displayed. Otherwise, this
should be an integer. For example, the following setting would always print the
5 most duplicated queries::

```python
QUERYCOUNT = {
    'DISPLAY_DUPLICATES': 5,
}
```

## License

This code is distributed under the terms of the [MIT license](LICENSE.txt).

## Testing

Run `python manage.py test querycount` to run the tests. Note that this will
modify your settings so that your project is in DEBUG mode for the duration
of the `querycount` tests.

(side-note: this project needs better tests; for the moment, there are only
smoke tests that set up the middleware and call two simple test views).

## Contributing

Bug fixes and new features are welcome! Fork this project and send a Pull Request
to have your work included. Be sure to add yourself to `AUTHORS.rst`.
````


HISTORY.md
# History

## 0.8.3

- Include querystring in host output ([PR #17](https://github.com/bradmontgomery/django-querycount/pull/17))

## 0.8.2

- Fixed statistics duplicate count.
- Fixed version (again).
- Fixed project url.

## 0.8.1

- Fixed internal version

## 0.8.0

- Improved duplicate detection caused by N+1 conditions.

## 0.7.0

- Introduced the [`RESPONSE_HEADER` Setting in #15](https://github.com/bradmontgomery/django-querycount/pull/15). Thanks @LifeCoder45!

## 0.6.0

- Fixed Issue #14 so that running tests doesn't result in `AttributeError: 'QueryCountMiddleware' object has no attribute 'get_response'`.
- Removed the `test_app` and replaced existing tests with very simple smoke tests.

## 0.5.0

- Added support for Django 1.10 Middleware.

## 0.4.2

- Fixed an issue that may cause an exception if we got a `None` value for the sql query.

## 0.4.1

- Fixed an issue where previous request's queries didn't get reset.
- Minor output formatting

## 0.4.0

- Included a count for duplicate sql queries. ([issue #7](https://github.com/bradmontgomery/django-querycount/issues/7))
- Included a new setting option, `DISPLAY_DUPLICATES` to control how many
  duplicate queries are printed in the shell.

## 0.3.0

- Added `IGNORE_SQL_PATTERNS`. [PR #6](https://github.com/bradmontgomery/django-querycount/pull/6). Thanks @GitFree
- Changed `IGNORE_PATTERNS` setting to `IGNORE_REQUEST_PATTERNS` (but still
  support the old setting)

## 0.2.0

- Added an `IGNORE_PATTERNS` setting and ability to ignore certain http requests.
- Slight project reorganization.

## 0.1.1

- Minor update to dynamically put version number in setup.py

## 0.1.0

- Added "elapsed" and "query count" log thresholds. [PR #4](https://github.com/bradmontgomery/django-querycount/pull/4). Thanks @alonisser

## 0.0.0

- Initial version. Huge thanks to @mrrrgn for making this look like a real project.


querycount/tests/test.py
from django.test import TestCase, modify_settings, override_settings
from querycount.middleware import QueryCountMiddleware


DEFAULT_QC_SETTINGS = {
    'THRESHOLDS': {
        'MEDIUM': 50,
        'HIGH': 200,
        'MIN_TIME_TO_LOG': 0,
        'MIN_QUERY_COUNT_TO_LOG': 0
    },
    'IGNORE_REQUEST_PATTERNS': [],
    'IGNORE_SQL_PATTERNS': [],
    'DISPLAY_DUPLICATES': None,
    'RESPONSE_HEADER': 'X-DjangoQueryCount-Count'
}

QC_MIDDLEWARE = {
    'append': 'querycount.middleware.QueryCountMiddleware',
}


@modify_settings(MIDDLEWARE=QC_MIDDLEWARE)
@override_settings(QUERYCOUNT=DEFAULT_QC_SETTINGS)
@override_settings(ROOT_URLCONF='querycount.tests.urls')
@override_settings(DEBUG=True)
class QueryCountTestCase(TestCase):
    # -------------------------------------------------------------------------
    # TODO: I want this app to be testable with just django's built-in tools,
    #       e.g. by running `python manage.py test querycount`.
    #
    #   - Figure out if there's a way to access the Middleware classes from a
    #     test client?
    #   - How to generate DB queries (read + write) reliably without jumping
    #     through the hoops of having a `test app` (that then needs to be
    #     monkey-patched in)
    #   - Should we just write a dump unit test for each of the
    #     QueryCountMiddleware's methods (probably).
    # -------------------------------------------------------------------------

    def setUp(self):
        self.querycount = QueryCountMiddleware()

    def test_empty(self):
        # Smoke test for a view that does no DB queries.
        resp = self.client.get("/empty/")
        self.assertEqual(resp.status_code, 200)

    def test_count(self):
        # Smoke test for a view that does a single DB queries.
        resp = self.client.get("/count/")
        self.assertEqual(resp.status_code, 200)
        self.assertEqual(int(resp['X-DjangoQueryCount-Count']), 1)

    def test_header_disabled(self):
        # Ensure removing the count header is effective
        disabled_settings = DEFAULT_QC_SETTINGS
        disabled_settings['RESPONSE_HEADER'] = None

        with self.settings(QUERYCOUNT=disabled_settings):
            resp = self.client.get("/count/")            
            self.assertEqual(resp.status_code, 200)
            self.assertFalse('X-DjangoQueryCount-Count' in resp)


querycount/tests/urls.py
from django.conf.urls import url
from . import views


urlpatterns = [
    url(r'^empty/$', views.empty, name='empty'),
    url(r'^count/$', views.count_migrations, name='count'),
]


querycount/tests/views.py
from django.db import connection
from django.http import HttpResponse


def empty(request):
    # An view with no DB queries.
    return HttpResponse("", content_type="text/plain")


def count_migrations(request):
    # A view that counts our existing migrations
    results = 0
    with connection.cursor() as cursor:
        cursor.execute("select count(*) from django_migrations")
        results = cursor.fetchone()[0]
    return HttpResponse(str(results), content_type="text/plain")


querycount/middleware.py
import re
import sys
import timeit
from collections import Counter
from textwrap import wrap

from django.conf import settings
from django.db import connections
from django.utils import termcolors

from .qc_settings import QC_SETTINGS

try:
    from django.utils.deprecation import MiddlewareMixin
except ImportError:
    MiddlewareMixin = object


class QueryCountMiddleware(MiddlewareMixin):
    """This middleware prints the number of database queries for each http
    request and response. This code is adapted from: http://goo.gl/UUKN0r.

    NOTE: This middleware is predominately written in the pre-Django 1.10 style,
    and uses the MiddlewareMixin for compatibility:

    https://docs.djangoproject.com/en/1.11/topics/http/middleware/#upgrading-pre-django-1-10-style-middleware

    """

    READ_QUERY_REGEX = re.compile(r"SELECT\s+.*")
    WHERE_CLAUSE_REGEX = re.compile(r"(\s+WHERE\s+)(.*)(\s+(?:GROUP|ORDER|HAVING)\s+)*")
    ID_REGEX = re.compile(r'(\s*[a-zA-Z_."]+?(?:_id|\."id)"\s*=\s*)\d+(\s*)')

    def __init__(self, *args, **kwargs):
        # Call super first, so the MiddlewareMixin's __init__ does its thing.
        super(QueryCountMiddleware, self).__init__(*args, **kwargs)

        if settings.DEBUG:
            self.request_path = None
            self.stats = {"request": {}, "response": {}}
            self.dbs = [c.alias for c in connections.all()]
            self.queries = Counter()
            self._reset_stats()

            self._start_time = None
            self._end_time = None
            self.host = None  # The HTTP_HOST pulled from the request

            # colorizing methods
            self.white = termcolors.make_style(opts=("bold",), fg="white")
            self.red = termcolors.make_style(opts=("bold",), fg="red")
            self.yellow = termcolors.make_style(opts=("bold",), fg="yellow")
            self.green = termcolors.make_style(fg="green")

            # query type detection regex
            # TODO: make stats classification regex more robust
            self.threshold = QC_SETTINGS["THRESHOLDS"]

    def _reset_stats(self):
        self.stats = {"request": {}, "response": {}}
        for alias in self.dbs:
            self.stats["request"][alias] = {"writes": 0, "reads": 0, "total": 0}
            self.stats["response"][alias] = {"writes": 0, "reads": 0, "total": 0}
        self.queries = Counter()

    def _count_queries(self, which):
        for c in connections.all():
            for q in c.queries:
                sql = q.get("sql")
                if not self._ignore_sql(q):
                    if sql and self.READ_QUERY_REGEX.search(sql) is not None:
                        self.stats[which][c.alias]["reads"] += 1
                    else:
                        self.stats[which][c.alias]["writes"] += 1
                    self.stats[which][c.alias]["total"] += 1

                    # Replace <some_id>=<some_number> with <some_id>=#number#
                    # in the WHERE clause to find duplicates that only
                    # vary by <some_number>, indicating N+1 condition.
                    where_clause_match = self.WHERE_CLAUSE_REGEX.search(sql)
                    if where_clause_match:
                        criteria_clause = self.ID_REGEX.sub(
                            r"\1#number#\2", where_clause_match.group(2)
                        )
                        repl = "r\1{}\3" if where_clause_match.group(3) else r"\1{}"
                        sql = self.WHERE_CLAUSE_REGEX.sub(
                            repl.format(criteria_clause), sql
                        )

                    self.queries[sql] += 1

            duplicates = sum(
                count for sql, count in self.queries.most_common() if count > 1
            )
            self.stats[which][c.alias]["duplicates"] = duplicates

    def _ignore_request(self, path):
        """Check to see if we should ignore the request."""
        return any(
            [
                re.match(pattern, path)
                for pattern in QC_SETTINGS["IGNORE_REQUEST_PATTERNS"]
            ]
        )

    def _ignore_sql(self, query):
        """Check to see if we should ignore the sql query."""
        return any(
            [
                re.search(pattern, query.get("sql"))
                for pattern in QC_SETTINGS["IGNORE_SQL_PATTERNS"]
            ]
        )

    def process_request(self, request):
        if settings.DEBUG and not self._ignore_request(request.path):
            self.host = request.META.get("HTTP_HOST", None)
            self.request_path = request.path
            self.query_string = request.META["QUERY_STRING"]
            self._start_time = timeit.default_timer()
            self._count_queries("request")

    def process_response(self, request, response):
        if settings.DEBUG and not self._ignore_request(request.path):
            self.request_path = request.path
            self._end_time = timeit.default_timer()
            self._count_queries("response")

            # Add query count header, if enabled
            if QC_SETTINGS["RESPONSE_HEADER"] is not None:
                response[QC_SETTINGS["RESPONSE_HEADER"]] = self._calculate_num_queries()

            self.print_num_queries()
            self._reset_stats()

        return response

    def _stats_table(self, which, path="", output=None):
        if output is None:
            if self.host:
                host_string = "http://{0}{1}".format(self.host, self.request_path)
            else:
                host_string = self.request_path

            if self.query_string:
                host_string += "?{0}".format(self.query_string)

            output = self.white("\n{0}\n".format(host_string))
            output += (
                "|------|-----------|----------|----------|----------|------------|\n"
            )
            output += (
                "| Type | Database  |   Reads  |  Writes  |  Totals  | Duplicates |\n"
            )
            output += (
                "|------|-----------|----------|----------|----------|------------|\n"
            )

        for db, stats in self.stats[which].items():
            if stats["total"] > 0:
                line = "|{w}|{db}|{reads}|{writes}|{total}|{duplicates}|\n".format(
                    w=which.upper()[:4].center(6),
                    db=db.center(11),
                    reads=str(stats["reads"]).center(10),
                    writes=str(stats["writes"]).center(10),
                    total=str(stats["total"]).center(10),
                    duplicates=str(stats["duplicates"]).center(12),
                )
                output += self._colorize(line, stats["total"])
                output += "|------|-----------|----------|----------|----------|------------|\n"
        return output

    def _duplicate_queries(self, output):
        """Appends the most common duplicate queries to the given output."""
        if QC_SETTINGS["DISPLAY_DUPLICATES"]:
            for query, count in self.queries.most_common(
                QC_SETTINGS["DISPLAY_DUPLICATES"]
            ):
                lines = ["\nExecuted {0} time(s).".format(count)]
                lines += wrap(query)
                lines = "\n".join(lines) + "\n"
                output += self._colorize(lines, count)
        return output

    def _totals(self, which):
        reads = 0
        writes = 0
        for db, stats in self.stats[which].items():
            reads += stats["reads"]
            writes += stats["writes"]
        return (reads, writes, reads + writes)

    def _colorize(self, output, metric):
        if metric > self.threshold["HIGH"]:
            output = self.red(output)
        elif metric > self.threshold["MEDIUM"]:
            output = self.yellow(output)
        else:
            output = self.green(output)
        return output

    def print_num_queries(self):
        # Request data
        output = self._stats_table("request")

        # Response data
        output = self._stats_table("response", output=output)

        # Summary of both
        if self._end_time and self._start_time:
            elapsed = self._end_time - self._start_time
        else:
            elapsed = 0

        count = self._calculate_num_queries()

        sum_output = "Total queries: {0} in {1:.4f}s \n\n".format(count, elapsed)
        sum_output = self._colorize(sum_output, count)
        sum_output = self._duplicate_queries(sum_output)

        # runserver just prints its output to sys.stderr, so we'll do that too.
        if (
            elapsed >= self.threshold["MIN_TIME_TO_LOG"]
            and count >= self.threshold["MIN_QUERY_COUNT_TO_LOG"]
        ):
            sys.stderr.write(output)
            sys.stderr.write(sum_output)

    def _calculate_num_queries(self):
        """
        Calculate the total number of request and response queries.
        Used for count header and count table.
        """
        request_totals = self._totals("request")
        response_totals = self._totals("response")

        return request_totals[2] + response_totals[2]  # sum total queries


querycount/qc_settings.py
"""
Default querycount settings.

"""

from django.conf import settings
from django.core.management.color import supports_color
from django.test.signals import setting_changed

QC_SETTINGS = {
    "IGNORE_REQUEST_PATTERNS": [],
    "IGNORE_SQL_PATTERNS": [],
    "THRESHOLDS": {
        "MEDIUM": 50,
        "HIGH": 200,
        "MIN_TIME_TO_LOG": 0,
        "MIN_QUERY_COUNT_TO_LOG": 0,
    },
    "DISPLAY_DUPLICATES": None,
    "RESPONSE_HEADER": "X-DjangoQueryCount-Count",
    "NOCOLOR": not supports_color(),
}


def _process_settings(**kwargs):
    """
    Apply user supplied settings.
    """

    # If we are in this method due to a signal, only reload for our settings
    setting_name = kwargs.get("setting", None)
    if setting_name is not None and setting_name != "QUERYCOUNT":
        return

    # Support the old-style settings
    if getattr(settings, "QUERYCOUNT_THRESHOLDS", False):
        QC_SETTINGS["THRESHOLDS"] = settings.QUERYCOUNT_THRESHOLDS

    # Apply new-style settings
    if not getattr(settings, "QUERYCOUNT", False):
        return

    # Duplicate display is a special case, configure it specifically
    if "DISPLAY_DUPLICATES" in settings.QUERYCOUNT:
        duplicate_settings = settings.QUERYCOUNT["DISPLAY_DUPLICATES"]
        if duplicate_settings is not None:
            duplicate_settings = int(duplicate_settings)
        QC_SETTINGS["DISPLAY_DUPLICATES"] = duplicate_settings

    # Apply the rest of the setting overrides
    for key in [
        "THRESHOLDS",
        "IGNORE_REQUEST_PATTERNS",
        "IGNORE_SQL_PATTERNS",
        "IGNORE_PATTERNS",
        "RESPONSE_HEADER",
        "NOCOLOR",
    ]:
        if key in settings.QUERYCOUNT:
            QC_SETTINGS[key] = settings.QUERYCOUNT[key]


# Perform initial load of settings
_process_settings()

# Subscribe to setting changes, via unit tests, etc
setting_changed.connect(_process_settings)
